### DISCUSSION TITLE
How to use `sklearn` regressors as `sktime` forecasters?

### DISCUSSION THREAD

**Discussion by IFV** [alan-turing-institute/sktime#2113] [Q&A]
**Describe the bug**

Dear all,

I am defining sklearn models as regressors. However,  the following msg still appears:

```ValueError: The estimator KNeighborsRegressor should be a Forecaster.``` 

Note clear to me the possible cause of error...
**To Reproduce**

```python
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor,AdaBoostRegressor

from sktime.forecasting.compose import StackingForecaster
from sktime.forecasting.naive import NaiveForecaster
import numpy as np
import pandas as pd

y_train = pd.Series(np.cos(np.arange(1,500)))

regressor0 = KNeighborsRegressor(n_neighbors=2)
regressor1 = RandomForestRegressor(random_state=0, n_estimators=100, max_depth=6)
regressor2 = AdaBoostRegressor(random_state=0, n_estimators=100)

forecasters = [("naive", NaiveForecaster()),
               ("KNN",regressor0),
               ("RF",regressor1),
               ("AdaBoost",regressor2)]

forecaster = StackingForecaster(forecasters=forecasters)
forecaster.fit(y=y_train, fh=[1])

y_pred = forecaster.predict()
```

**Expected behavior**

**Additional context**

**Versions**
<details>

</details>

System:
    python: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0]
executable: /home/...
   machine: Linux-5.4.0-70-generic-x86_64-with-glibc2.31

Python dependencies:
          pip: 21.2.4
   setuptools: 58.0.4
      sklearn: 1.0.2
       sktime: 0.10.0
  statsmodels: 0.13.2
        numpy: 1.21.5
        scipy: 1.7.3
       pandas: 1.3.5
   matplotlib: 3.5.0
       joblib: 1.1.0
        numba: None
     pmdarima: 1.8.4
      tsfresh: None

**Comment by fkiraly** [comment]
@IFV, this is not a bug - the forecasters passed to the `StackingForecaster` need to be forecasters, i.e., following the `sktime` `BaseForecaster` interface.

If you want to use `sklearn` regressors, that is possible - you just need to say *how* you want to use them as regressors, since there are multiple ways. The most direct way would be to use `make_reduction` and specify a reduction algorithm (= a way to turn a regressor into a forecaster).

See the [forecasting tutorial](https://www.sktime.org/en/stable/examples/01_forecasting.html), section 3, for reduction and other composition patterns.

You could do sth like this:

```python
....
from sktime.forecasting.compose import make_reduction

regressor0 = KNeighborsRegressor(n_neighbors=2)
forecaster0 = make_reduction(regressor0, window_length=15, strategy="recursive")
...

forecasters = [("naive", NaiveForecaster()),
               ("KNN", forecaster0),
    ...
 
forecaster = StackingForecaster(forecasters=forecasters)
forecaster.fit(y=y_train, fh=[1])

y_pred = forecaster.predict()
```
(fill in the dots using the same pattern)

You have various choices for the reduction, e.g., the window length or the reduction algorithm (via `strategy` arg).

**Reply by fkiraly** [reply]
(converted to discussion since not a bug, and it might be generally helpful to point this out)

**Reply by IFV** [reply]
I see. Many thanks!

**Reply by fkiraly** [reply]
PS: you could also use stacking *first*, and then apply reduction *second*. That way, you have only a single set of reduction parameters. Different algorithm, but similar idea. Stacking would be for sklearn regressors I.e., `make_reduction(Stacking(forecasters))`, where you need to find a stacking algorithm compatible with `sklearn` regressors.
